### PR TITLE
Malf AI Rework

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports over @KorPhaeron's rework/retooling of Malf AI: https://github.com/tgstation/tgstation/pull/13759 along with all balance changes and fixes

Malf AI is an interesting gamemode, but it suffers from a number of problems that have plagued it for a long time. This PR attempts to address a large number of these concerns and overall make the concept of a malfunctioning AI far more interesting.

Some of the big problems with malf AI is that it's a super easy game mode to meta, there's little to no team-work, the AI has incredibly heavy dependency on borgs, and the mode has very binary gameplay. Futhermore, rounds tends to play out the same--AI hacks for a long time, only to be bum-rushed during the last 10 minutes in a mad-dash assault on the core.

What this PR does:

Malfunction as a *distinct* game-mode has been removed. Instead, the functionality of malf AI has been merged into ***traitor*** AI's. This means that if an AI is a traitor, it's also malfunctioning.

A traitor AI has these abilities:
- The ability to hack APCs
- The ability to purchase malf modules (like hostile lockdown, override machine, machine overload, etc)
 - Naturally the AI would have the powers of any modules it purchases
- Ability to shunt
- A syndicate encryption key in its headset

The above, alone, resolves a number of issues with both malfunction, as a game mode, and traitor AIs. It makes it a bit harder "meta" the type of round it is, it heavily reduces reliance on borgs (as you have your modules to purchase), and it allows you to co-ordinate with traitors, if you want. You could set up grand schemes like having a traitor go around emagging a bunch of APCs for plausible deniability, all the while you slowly start hacking others

What about blowing up the station? 

This still exists in the form of a new malf module called the Doomsday Device. When purchased (130 CPU) and activated it initiates Delta and sets a timer for 7.5 minutes. If the AI is killed, in that time frame, then the gamma alert drops down to red and the crew is safe---the round will not end if the AI is killed. If the doomsday device is *not* disarmed after 7.5 minutes, then it will detonate, killing all living mobs on the station and ending the round.

But why would traitors and a malfunctioning AI work together if it's just going to kill them all with the Doomsday device?

There's special exception for the "escape alive" objective. If the person survives the Doomsday Device (by being off station), they will count has having "escaped alive". In this way, traitors can work towards their objectives and work with the AI and not really care about *having* to be on the shuttle if the round ends early; just to survive until the end and complete their other two objectives.

Malf/Traitor AI effectively being the same thing also resolves two other issues:

People can turn on the malf AI preference and never worry about rolling for AI, but still end up as a malf AI. Also the fact that it's a distinct game mode inherently makes it fairly rare to crop up, while malf AI being a part of traitor inherently means more chances for it to crop up and in far more interesting and dynamic situations.

Quality of Life Changes
- Hacking an APC now gives an alert when its hacking. Clicking on the alert will move your vision to that APC. When the APC is finished hacking the alert will clear and a ding sound will be played

Other Related Things
- Adds in AI upgrade modules, of which there are two: Combat Module and Lip Reading module. Both of these are currently unavailable, but they could be made makeable by R&D later or, better yet, as loot in some place like ~~lavaland~~Deep Space :wink: 

Fixes
- Fixes not being able to assign traitor to non-humans or remove it from them

Code Related
- Pinpointers use their own list instead of `in world`
- Less damn snowflake all over for one game mode

:cl: KorPhaeron and Fox McCloud
tweak: Merges Malf AI into traitor AI
tweak: Malf AI's starting CPU reduced from 100 to 50
tweak: Hacking APCs gives 10 additional CPU
add: Hacking APCs throws an alert to the AI when it's hacking that automatically clears when hacking is complete (sounds too)
add: Traitor AI's start off with a syndicate headset
add: Adds doomsday device module for AI's for 130 CPU; when activated, it will purge the station of all life if the AI is not destroyed after 7.5 minutes
add: Adds in AI upgrade modules which allows you to giving hacking abilities to the AI or allow it to have increase surveillance capabilities
/:cl:

### CODE REVIEW

File: code/game/gamemodes/malfunction/Malf_Modules.dm [hunk 1]
```diff
@@ -16,6 +16,103 @@
...
+	if(!timing)
+		fast_processing -= src
+		return
+	var/sec_left = seconds_remaining()
+	if(sec_left <= 0)
+		timing = 0
+		detonate(T.z)
+		qdel(src)
+	else
+		if(!(sec_left % 60) && (!milestones[sec_left]))
```
Comment (Fox-McCloud) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: I'll probably need help with this, as this runtimes. Bad index.

Comment (KorPhaeron) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: revert to the old timer that worked properly

File: code/modules/power/apc.dm [hunk 2]
```diff
@@ -1024,70 +1001,72 @@
...
-				for(var/datum/mind/AI_mind in ticker.mode.malf_ai)
-					var/mob/living/silicon/ai/A = AI_mind.current // the current mob the mind owns
-					if(A.stat != DEAD)
-						point.the_disk = A //The pinpointer tracks the AI back into its core.
+	if(occupier.parent && occupier.parent.stat != DEAD)
+		occupier.mind.transfer_to(occupier.parent)
+		occupier.parent.shunted = 0
+		occupier.parent.adjustOxyLoss(occupier.getOxyLoss())
+		occupier.parent.cancel_camera()
+		qdel(occupier)
```
Comment (Fox-McCloud) [inline code/modules/power/apc.dm]: Hmm, my latest commit didn't go through; need to add the pinpointer to this, again.

Comment (Fox-McCloud) [inline code/modules/power/apc.dm]: Re-add the disk pointing to the core.

**Comment by Aurorablade** [comment]
okay while i like the changes..i am not sure if i agree with removing the game mode being the solution here..but thats me.

**Comment by Fox-McCloud** [comment]
With these changes, having it be a specific mode is...redundant. It's like having traitor, with only one traitor ever being picked. Wouldn't be nearly as dynamic or as fun.

File: code/game/machinery/computer/communications.dm [hunk 3]
```diff
@@ -514,7 +514,7 @@
 		if(!shuttlecaller.stat && shuttlecaller.client && istype(shuttlecaller.loc,/turf))
 			return ..()
 
-	if(ticker.mode.name == "revolution" || ticker.mode.name == "AI malfunction" || sent_strike_team)
+	if(ticker.mode.name == "revolution" || sent_strike_team)
```
Comment (tigercat2000) [inline code/game/machinery/computer/communications.dm]: `if(GAMEMODE_IS_REVOLUTION || sent_strike_team)`

File: code/game/gamemodes/malfunction/Malf_Modules.dm (hunk 1 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/malfunction/Malf_Modules.dm]: went with a small internal cooldown.

**Comment by tigercat2000** [comment]
I don't really like this, especially given that traitor AI's won't get an objective to nuke the station. Doing so anyways seems dangerously close to things like releasing the singularity under the rules.

**Comment by Fox-McCloud** [comment]
Given the leeway traitor AI's are already given (and how hostile they become once discovered), not to mention their law specifically grants them this "**accomplish your objective at all costs**" is the specific law they get, well; they're already granted the authority to do somethin' like that.

**Comment by Aurorablade** [comment]
i still like the game mode :/ i will just be in my box...

**Comment by KingPhilipIII** [comment]
I could go either way. Been playing a few rounds on TG station and I notice that malf AIs crop up a lot more with the fact that traitor AIs are automatically malfunctioning. I think most of my reservations come from a usual resistance to change, because I have no real qualms about what this PR does.

**Comment by BiberDark** [comment]
I Like this but , is it stil like in the past that the traitor ai cant be resseted and is this visible to others ?

**Comment by KorPhaeron** [comment]
my grand vision spreads

**Pull request merged**